File: LeetCode/1438. Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/1438.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int LongestSubarray(int[] nums, int limit) {
        // LinkedLists to store the indices of the minimum and maximum values in the window
        LinkedList<int> maxDeque = new LinkedList<int>();
        LinkedList<int> minDeque = new LinkedList<int>();
        int start = 0, maxLength = 0;

        for (int end = 0; end < nums.Length; end++) {
            // Maintain the maxDeque
            while (maxDeque.Count > 0 && nums[maxDeque.Last.Value] <= nums[end]) {
                maxDeque.RemoveLast();
            }
            maxDeque.AddLast(end);

            // Maintain the minDeque
            while (minDeque.Count > 0 && nums[minDeque.Last.Value] >= nums[end]) {
                minDeque.RemoveLast();
            }
            minDeque.AddLast(end);

            // Ensure the current window satisfies the condition
            while (nums[maxDeque.First.Value] - nums[minDeque.First.Value] > limit) {
                if (maxDeque.First.Value == start) {
                    maxDeque.RemoveFirst();
                }
                if (minDeque.First.Value == start) {
                    minDeque.RemoveFirst();
                }
                start++;
            }

            // Update the maximum length of the subarray
            maxLength = Math.Max(maxLength, end - start + 1);
        }

        return maxLength; // Return the size of the longest subarray
    }

    public static void Main(string[] args) {
        var solution = new Solution();
        int[] nums = {10, 1, 2, 4, 7, 2};
        int limit = 5;
        Console.WriteLine(solution.LongestSubarray(nums, limit)); // Expected output: 4
    }
}
//other people's solutions
/*
public class Solution {
    public int LongestSubarray(int[] nums, int limit) {
        Dequeue minQueue = new Dequeue();
        Dequeue maxQueue = new Dequeue();
        int ans = 0;
        int j = 0;
        for(int i = 0; i < nums.Length; i++) {
            while(minQueue.Count() > 0 && nums[minQueue.PeekBack()] >= nums[i]) minQueue.DequeueBack();
            minQueue.EnqueueBack(i);
            
            while(maxQueue.Count() > 0 && nums[maxQueue.PeekBack()] <= nums[i]) maxQueue.DequeueBack();
            maxQueue.EnqueueBack(i);
            
            if(nums[maxQueue.PeekFront()] - nums[minQueue.PeekFront()] > limit) {
                j++;
                
                if(j > minQueue.PeekFront()) minQueue.DequeueFront();
                if(j > maxQueue.PeekFront()) maxQueue.DequeueFront();
            }
            else {
                ans = Math.Max(ans, i - j + 1);
            }
        }
        
        return ans;
    }
}

public class Dequeue {
    int idx;
    List<int> nums;
    
    public Dequeue() {
        idx = 0;
        nums = new List<int>();
    }
    
    public void EnqueueBack(int num) {
        nums.Add(num);
    }
    
    public void DequeueBack() {
        nums.RemoveAt(nums.Count-1);
    }
    
    public void DequeueFront() {
        idx++;
    }
    
    public int Count() {
        return nums.Count - idx;
    }
    
    public int PeekBack() {
        return nums[^1];
    }
    
    public int PeekFront() {
        return nums[idx];
    }
}
*/

/*
public class Solution {
    public int LongestSubarray(int[] nums, int limit) {
        // TODO: 2 deque
        // use deque where it stores indexes, first item is largest
        // each iteration, compares with the largest so far
        // if nums[i] - deque[0] > limit, maxLength=0
        // else maxLength++
        var minDeque = new LinkedList<int>();
        var maxDeque = new LinkedList<int>();
        var curMaxLength = 0;
        var maxLength = 0;
        // [8,2,4,7]
        //      i
        // deque = [0,2]
        // limit = 4
        // diff = 8-2
        // curMaxLength = 1
        // maxLength = 1
        int i=0;
        for(int j=0;j<nums.Length;j++){
            while(maxDeque.Count>0 && nums[maxDeque.Last.Value] < nums[j]){
                maxDeque.RemoveLast();
            }
            maxDeque.AddLast(j);
            while(minDeque.Count>0 && nums[minDeque.Last.Value] > nums[j]){
                minDeque.RemoveLast();
            }
            minDeque.AddLast(j);

            while(nums[maxDeque.First.Value] - nums[minDeque.First.Value] > limit){
                if(maxDeque.First.Value == i) {
                    maxDeque.RemoveFirst();
                }
                if(minDeque.First.Value == i) {
                    minDeque.RemoveFirst();
                }
                i++; // shrink right
            }
            maxLength = Math.Max(maxLength, j - i + 1);
        }
        return maxLength;
    }
}
*/


/*
public class Solution {
   private class Deq
{
    private int[] store;
    private int length;
    private int head;
    private int _size;

    public Deq(int size)
    {
        store = new int[size];
        _size = size;
    }

    public void SetFront(int val)
    {
        store[0] = val;
        head = 0;
        length = 1;
    }

    public void AddBack(int val)
    {
        length++;
        store[(head + length - 1)%_size] = val;
    }

    public int PeekFront()
    {
        if (length == 0)
        {
            throw new Exception();
        }

        return store[head%_size];
    }

    public int PopFront()
    {
        if (length == 0)
        {
            throw new Exception();
        }

        head++;
        length--;
        return store[(head - 1)%_size];
    }

    public int PeekBack()
    {
        if (length == 0)
        {
            throw new Exception();
        }

        return store[(head + length - 1)%_size];
    }

    public int PopBack()
    {
        if (length == 0)
        {
            throw new Exception();
        }

        length--;
        return store[(head + length)%_size];
    }

    public int Count()
    {
        return length;
    }
}

public int LongestSubarray(int[] nums, int limit)
{
    var max = new Deq(nums.Length);
    var min = new Deq(nums.Length);
    var i = 0;
    var j = 0;
    var result = 1;
    while (i < nums.Length && j < nums.Length)
    {
        if (max.Count() == 0 || nums[j] >= nums[max.PeekFront()] )
        {
            max.SetFront(j);
        }
        else
        {
            while (nums[max.PeekBack()] < nums[j])
            {
                max.PopBack();
            }

            max.AddBack(j);
        }

        if (min.Count() == 0 || nums[j] <= nums[min.PeekFront()])
        {
            min.SetFront(j);
        }
        else
        {
            while (nums[min.PeekBack()] > nums[j])
            {
                min.PopBack();
            }

            min.AddBack(j);
        }
        if (Math.Abs(nums[max.PeekFront()] - nums[min.PeekFront()]) <= limit)
        {
            if (j - i + 1 > result)
            {
                result = j - i + 1;
            }
        }
        else
        {
            while (Math.Abs(nums[max.PeekFront()] - nums[min.PeekFront()]) > limit)
            {
                var d = max.PeekFront() < min.PeekFront() ? max : min;
                i = d.PopFront() + 1;
            }
        }

        j++;
    }

    return result;
}
}
*/

File: LeetCode/1438. Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/1438.java
import java.util.Deque;
import java.util.LinkedList;

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        // Deques to store the indices of the minimum and maximum values in the window
        Deque<Integer> maxDeque = new LinkedList<>();
        Deque<Integer> minDeque = new LinkedList<>();
        int start = 0, maxLength = 0;

        for (int end = 0; end < nums.length; end++) {
            // Maintain the maxDeque and minDeque
            while (!maxDeque.isEmpty() && nums[maxDeque.peekLast()] <= nums[end]) {
                maxDeque.pollLast();
            }
            while (!minDeque.isEmpty() && nums[minDeque.peekLast()] >= nums[end]) {
                minDeque.pollLast();
            }
            maxDeque.offerLast(end);
            minDeque.offerLast(end);

            // Ensure the current window satisfies the condition
            while (nums[maxDeque.peekFirst()] - nums[minDeque.peekFirst()] > limit) {
                if (maxDeque.peekFirst() == start) {
                    maxDeque.pollFirst();
                }
                if (minDeque.peekFirst() == start) {
                    minDeque.pollFirst();
                }
                start++;
            }

            // Update the maximum length of the subarray
            maxLength = Math.max(maxLength, end - start + 1);
        }

        return maxLength; // Return the size of the longest subarray
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {10, 1, 2, 4, 7, 2};
        int limit = 5;
        System.out.println(solution.longestSubarray(nums, limit)); // Expected output: 4
    }
}

//other people's solution

/*
 class Solution {

    // maitain max val of cur slide window in head of deque
    static int[] maxDeque ;
    static int[] minDeque;
    static int maxHead;
    static int maxTail;
    static int minHead;
    static int minTail;

    // maintain min val of cur silde window in head of deque
    public int longestSubarray(int[] nums, int limit) {
        int n = nums.length;
        maxDeque = new int[n];
        minDeque = new int[n];

        // queue range [h,t)
        maxHead=0;
        maxTail=0;
        minHead=0;
        minTail=0;

        // take each num in nums as start of slide window
        // get max len of window with any two elements in window is less than or equal to limit.

        int res =0;
        int r=0;
        for(int l =0;l<n;l++){

            // window is [l,r), check if add r to window still meet condition
            while(r<n&&withinLimit(nums,nums[r],limit)){
                addToLast(nums,r);
                r++;
            }
            // [l,r) is max len of a valid window using l as start 
            res= Math.max(res, r-l);

           removeExpired(l);
        }

        return res;
        
    }

    public boolean withinLimit(int[] nums,int num,int limit){
        int max = maxHead<maxTail?Math.max(nums[maxDeque[maxHead]], num):num;
        int min = minHead<minTail?Math.min(nums[minDeque[minHead]],num):num;

        return max-min<=limit;
    }

    public void addToLast(int[] nums,int r){
        // add to max deque 
        while(maxHead<maxTail&&nums[maxDeque[maxTail-1]]<=nums[r]){
             maxTail--;
        }
        maxDeque[maxTail++]=r;

        // add to min deque
        while(minHead<minTail&&nums[minDeque[minTail-1]]>=nums[r]){
            minTail--;
        }
        minDeque[minTail++]=r;

    }

    public void removeExpired(int l){
        if(maxDeque[maxHead]==l){
            maxHead++;
        }

        if(minDeque[minHead]==l){
            minHead++;
        }
    }

}
 */

 /*
  class Solution {
    class MonotonicQueue{
        LinkedList<Integer> maxq = new LinkedList<>();
        LinkedList<Integer> minq = new LinkedList<>();

        public void push(int val) {
            while (!maxq.isEmpty() && maxq.getLast() < val) {
                maxq.removeLast();
            }
            maxq.addLast(val);

            while (!minq.isEmpty() && minq.getLast() > val) {
                minq.removeLast();
            }
            minq.addLast(val);
        }

        public int max(){
            return maxq.getFirst();
        }
        public int min() {
            return minq.getFirst();
        }

        public void pop(int val) {
            if (maxq.getFirst() == val) {
                maxq.removeFirst();
            }
            if (minq.getFirst() == val) {
                minq.removeFirst();
            }
        }
    }
    public int longestSubarray(int[] nums, int limit) {
        int i=0,j= 0;
        int len = 0, maxLen = 0;
        MonotonicQueue window = new MonotonicQueue();
        while (j<nums.length) {
            window.push(nums[j]);
            j++;
            len++;
            while (window.max()-window.min() > limit) {
                window.pop(nums[i]);
                i++;
                len--;
            }
            // len = j-i
            maxLen = Math.max(maxLen,len);
        }
        return maxLen;
    }
}

  */

  /*
   class Solution {
    public int longestSubarray(int[] nums, int limit) {
        LinkedList<Integer> minq = new LinkedList<>();
        LinkedList<Integer> maxq = new LinkedList<>();
        int res = 0, i = 0;
        for(int j = 0; j < nums.length; j++) {
            while(!minq.isEmpty() && minq.peekLast() > nums[j]) {
                minq.removeLast();
            }
            minq.addLast(nums[j]);
            while(!maxq.isEmpty() && maxq.peekLast() < nums[j]) {
                maxq.removeLast();
            }
            maxq.addLast(nums[j]);
            while(maxq.peekFirst() - minq.peekFirst() > limit) {
                if(maxq.peekFirst() == nums[i]) {
                    maxq.removeFirst();
                }
                if(minq.peekFirst() == nums[i]) {
                    minq.removeFirst();
                }
                i++;
            }
            res = Math.max(res, j - i + 1);
        }
        return res;
    }
}
   */
  /*
   class Solution {
    private static class Pair implements Comparable<Pair> {
        int index;
        int value;
        private Pair(int index, int value) {
            this.index = index;
            this.value = value;
        }

        @Override
        public int compareTo(Pair o) {
            if (this.value == o.value) {
                return this.index - o.index;
            }
            return this.value - o.value;
        }
    }

    public int longestSubarray(int[] nums, int limit) {
        int low = 0, high = 0, answer = 0;
        PriorityQueue<Integer> min = new PriorityQueue<>();
        PriorityQueue<Integer> max = new PriorityQueue<>(Collections.reverseOrder());
        while (high < nums.length) {
            min.add(nums[high]);
            max.add(nums[high]);
            if (max.peek() - min.peek() > limit) {
                answer = Math.max(answer, high - low);
                while (max.peek() - min.peek() > limit) {
                    min.remove(nums[low]);
                    max.remove(nums[low]);
                    low++;
                }
            }
            high++;
        }
        return Math.max(answer, high - low);
    }
}
   */